### ISSUE TITLE
Overlapping Mask Generation and LLM Inference

### ISSUE THREAD

**Issue by pmfirestone** [mlc-ai/xgrammar#112]
In section 3.5 of the paper you describe how overlapping the generation of the mask and inference of the LLM could save time. Is this something you have implemented, or only a theoretical possibility? If you have implemented it already, where could I find more information about how you did it?

**Comment by Ubospica** [comment]
Yes, we have already implemented that in open-source LLM serving systems:

In MLC-LLM (a C++ serving engine),
- First, GPU kernels of LLM inference are called but not synchronized;
- Then we call mask generation at https://github.com/mlc-ai/mlc-llm/blob/main/cpp/serve/logit_processor.cc#L416. This code is synchronously executed on the CPU, but the GPU execution is running asynchronously at the same time.
- Then we synchronize the CPU and GPU at a later time when we need to copy the results of GPU sampling to the CPU and print the output. The code is at https://github.com/mlc-ai/mlc-llm/blob/d23d6f51e4d72cb5105ae79ecb1e854ca490acb7/cpp/serve/sampler/gpu_sampler.cc#L685.

In SGLang (a Python serving engine), the overlapping is achieved by the scheduler which maintains a CPU thread and a GPU thread. XGrammar's logic is running on the CPU. The two threads are synchronized when necessary. Check out https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/constrained/xgrammar_backend.py and https://github.com/sgl-project/sglang/pull/2095.

Our benchmark results in the blog post are after overlapping.

**Comment by pmfirestone** [comment]
Thank you for the reply! I appreciate the prompt communication. Congratulations on integrating XGrammar into these serving engines.
